src/Plutarch/Test/QuickCheck.hs

```diff
@@ -0,0 +1,265 @@
+
+{- | Converts Plutarch Functions into `Testable` Haskell function of TestableTerms.
+
+ @since 2.0.0
+-}
+fromPFun ::
+    forall a b c.
+    ( c ~ PTT (a :--> b)
+    , FromPFunN a b c
+    ) =>
+    (forall s. Term s (a :--> b)) ->
+    TestableFun c (a :--> b)
+fromPFun = fromPFun' (Proxy @(PTT (a :--> b)))
+
+{- | Extracts all @TestableTerm@s from give Plutarch function.
+
+ @since 2.0.0
+-}
+type family PLamArgs (p :: S -> Type) :: [Type] where
+    PLamArgs (a :--> b) = TestableTerm a : PLamArgs b
+    PLamArgs _ = '[]
+
+{- | Make property by directly comparing behavior of Plutarch function
+     to Haskell counterpart.  This function will expect all Plutarch
+     types to be `plift`able and `pshow`able.  With given TestableTerm
+     generator, it will generate value for each arguments. Then, it
+     will apply generated value and lifted value to Plutarch and
+     haskell function. Once all arguments are applied, It will check
+     the equality of results.
+
+ @since 2.0.0
+-}
+class (PLamArgs p ~ args) => HaskEquiv (h :: Type) (p :: S -> Type) args where
+    haskEquiv :: h -> TestableTerm p -> NP Gen args -> Property
+
+-- TODO: Shrinking support
+
+-- | @since 2.0.0
+instance
+    forall ha hb pa pb hbArgs.
+    ( PLamArgs pb ~ hbArgs
+    , HaskEquiv hb pb hbArgs
+    , PLifted pa ~ ha
+    , PLift pa
+    , PShow pa
+    ) =>
+    HaskEquiv (ha -> hb) (pa :--> pb) (TestableTerm pa ': hbArgs)
+    where
+    haskEquiv h (TestableTerm p) (g :* gs) =
+        forAll g $ \(TestableTerm x) -> haskEquiv (h $ plift x) (TestableTerm $ p # x) gs
+
+-- | @since 2.0.0
+instance (PLamArgs p ~ '[], PLift p, PLifted p ~ h, Eq h) => HaskEquiv h p '[] where
+    haskEquiv h (TestableTerm p) _ = property $ plift p == h
+
+{- | Simplified version of `haskEquiv`. It will use arbitrary instead of
+     asking custom generators.
+
+ @since 2.0.0
+-}
+haskEquiv' ::
+    forall h p args.
+    ( PLamArgs p ~ args
+    , HaskEquiv h p args
+    , All Arbitrary args
+    ) =>
+    h ->
+    (forall s. Term s p) ->
+    Property
+haskEquiv' h p = haskEquiv h (TestableTerm p) $ hcpure (Proxy @Arbitrary) arbitrary
+
+{- | Placeholder for a polymorphic type. Plutarch equivalence of QuickCheck's `A`.
+
+ @since 2.0.0
+-}
+type PA :: S -> Type
+type PA = PInteger
+
+{- | Same as `PA`.
+
+ @since 2.0.0
+-}
+type PB :: S -> Type
+type PB = PInteger
+
+{- | Same as `PA`.
+
+ @since 2.0.0
+-}
+type PC :: S -> Type
+type PC = PInteger
+
+{- | This shinker "simplifies" underlaying plutarch representation. When
+     shrinking List, this shinker is always preferable.
+
+ @since 2.0.0
+-}
+shrinkPLift ::
+    forall a.
+    ( PLift a
+    , Arbitrary (PLifted a)
+    ) =>
+    TestableTerm a ->
+    [TestableTerm a]
+shrinkPLift = fmap pconstantT . shrink . pliftT
+
+{- | This generator uses `Arbitrary` instance of Haskell representation to
+     make value and lift that into Plutarch.
+
+ @since 2.0.0
+-}
```

Comment:
"uses the 'Arbitrary' instance of a Haskell representation'
'make a value and lift it'